File: 3/day3.py
# https://adventofcode.com/2024/day/3
import re
                    
def day3(text: str) -> None:
    def part1() -> None:
        pat1 = r"mul\((\d+),(\d+)\)"
        matches = re.findall(pat1, text)
        print(sum(int(num[0]) * int(num[1]) for num in matches))

    def part2() -> None:    
        pat2 = re.findall(r"(?:mul\((\d+),(\d+)\))|(do\(\)|don't\(\))", text)
        enabled = True
        ans = 0
        for match in pat2:
            if match[2] == "" and enabled:
                ans += int(match[0]) * int(match[1])
            else:
                enabled = match[2] == "do()"
        print(ans)
    
    part1()
    part2()
    
def main(file_path: str):
    with open("./3/input.txt") as fin:
        s = fin.read().strip()
    day3(s)
    
if __name__ == '__main__':
    # main('./3/input0.txt')
    main('./3/input.txt')

"""
Part 1:
input.txt: 178886550

Part 2:
input.txt: 87163705
""" 

File: 3/day3.js
const fs = require('fs');

function getData(filePath) {
    let text = '';
    try {
      const data = fs.readFileSync(filePath, 'utf8');
      text = data.trim();
    } catch (err) {
      console.error("Error reading file:", err);
    }
    return text;
}

function part1(filePath) {
    const pattern = /mul\((\d+),(\d+)\)/g;
    const matches = getData(filePath).matchAll(pattern);
    /*
    This block of codes can be rewritten using functional programming style below
    
    let sum = 0;

    for (const match of matches) {
        sum += parseInt(match[1], 10) * parseInt(match[2], 10);
    }
    */
    const sum = [...matches]
        .map(match => parseInt(match[1], 10) * parseInt(match[2], 10))
        .reduce((acc, val) => acc + val, 0);

    console.log(sum);
}

function part2(filePath) {
    const text = getData(filePath);
    const pattern = /(?:mul\((\d+),(\d+)\))|(do\(\)|don't\(\))/g;
    const matches = [...text.matchAll(pattern)]; // Use matchAll for iterable matches
    /*
    This block of codes can be rewrittend using the FP style below:
    let enabled = true;
    let ans = 0;

    for (const match of matches) {
        if (match[3] === undefined && enabled) {
            // Process "mul" matches
            ans += parseInt(match[1], 10) * parseInt(match[2], 10);
        } else {
            // Process "do()" or "don't()" matches
            enabled = match[3] === "do()";
        }
    }
    console.log(ans);
    */
    const result = matches.reduce((acc, match) => {
        if (match[3] === undefined && acc.enabled) {
          // Process "mul" matches
          return {
            enabled: true,
            sum: acc.sum + parseInt(match[1], 10) * parseInt(match[2], 10)
          };
        } else {
          // Process "do()" or "don't()" matches
          return {
            enabled: match[3] === "do()",
            sum: acc.sum
          };
        }
      }, { enabled: true, sum: 0 });
    
      console.log(result.sum);
}

part1('./3/input.txt')
part2('./3/input.txt')